<!DOCTYPE HTML>
<html lang="en">
    <head>
        <meta charset="UTF-8">
        <title>Python Exception - PyO3 user guide</title>
        <meta content="text/html; charset=utf-8" http-equiv="Content-Type">
        <meta name="description" content="PyO3 user guide">
        <meta name="viewport" content="width=device-width, initial-scale=1">

        <base href="">

        <link rel="stylesheet" href="book.css">
        <link href="https://fonts.googleapis.com/css?family=Open+Sans:300italic,400italic,600italic,700italic,800italic,400,300,600,700,800" rel="stylesheet" type="text/css">
        <link href="https://fonts.googleapis.com/css?family=Source+Code+Pro:500" rel="stylesheet" type="text/css">

        <link rel="shortcut icon" href="favicon.png">

        <!-- Font Awesome -->
        <link rel="stylesheet" href="https://maxcdn.bootstrapcdn.com/font-awesome/4.3.0/css/font-awesome.min.css">

        <link rel="stylesheet" href="highlight.css">
        <link rel="stylesheet" href="tomorrow-night.css">
        <link rel="stylesheet" href="ayu-highlight.css">

        <!-- Custom theme -->
        

        

        <!-- Fetch Clipboard.js from CDN but have a local fallback -->
        <script src="https://cdn.jsdelivr.net/clipboard.js/1.6.1/clipboard.min.js"></script>
        <script>
            if (typeof Clipboard == 'undefined') {
                document.write(unescape("%3Cscript src='clipboard.min.js'%3E%3C/script%3E"));
            }
        </script>

        <!-- Fetch JQuery from CDN but have a local fallback -->
        <script src="https://code.jquery.com/jquery-2.1.4.min.js"></script>
        <script>
            if (typeof jQuery == 'undefined') {
                document.write(unescape("%3Cscript src='jquery.js'%3E%3C/script%3E"));
            }
        </script>

        <!-- Fetch store.js from local - TODO add CDN when 2.x.x is available on cdnjs -->
        <script src="store.js"></script>

        <!-- Custom JS script -->
        

    </head>
    <body class="light">
        <!-- Set the theme before any content is loaded, prevents flash -->
        <script type="text/javascript">
            var theme = store.get('mdbook-theme');
            if (theme === null || theme === undefined) { theme = 'light'; }
            $('body').removeClass().addClass(theme);
        </script>

        <!-- Hide / unhide sidebar before it is displayed -->
        <script type="text/javascript">
            var sidebar = store.get('mdbook-sidebar');
            if (sidebar === "hidden") { $("html").addClass("sidebar-hidden") }
            else if (sidebar === "visible") { $("html").addClass("sidebar-visible") }
        </script>

        <div id="sidebar" class="sidebar">
            <ul class="chapter"><li class="affix"><a href="./overview.html">Overview</a></li><li class="affix"><a href="./getting_started.html">Getting Started</a></li><li><a href="./conversions.html"><strong>1.</strong> Type Conversions</a></li><li><a href="./exception.html" class="active"><strong>2.</strong> Python Exception</a></li><li><a href="./module.html"><strong>3.</strong> Python Module</a></li><li><a href="./function.html"><strong>4.</strong> Python Function</a></li><li><a href="./class.html"><strong>5.</strong> Python Class</a></li><li><a href="./parallelism.html"><strong>6.</strong> Parallelism</a></li><li><a href="./distribution.html"><strong>7.</strong> Distribution</a></li></ul>
        </div>

        <div id="page-wrapper" class="page-wrapper">

            <div class="page" tabindex="-1">
                <div id="menu-bar" class="menu-bar">
                    <div class="left-buttons">
                        <i id="sidebar-toggle" class="fa fa-bars"></i>
                        <i id="theme-toggle" class="fa fa-paint-brush"></i>
                    </div>

                    <h1 class="menu-title">PyO3 user guide</h1>

                    <div class="right-buttons">
                        <a href="print.html">
                            <i id="print-button" class="fa fa-print" title="Print this book"></i>
                        </a>
                    </div>
                </div>

                <div id="content" class="content">
                    <a class="header" href="./exception.html#python-exception" id="python-exception"><h1>Python Exception</h1></a>
<a class="header" href="./exception.html#define-a-new-exception" id="define-a-new-exception"><h2>Define a new exception</h2></a>
<p>You can use the <code>py_exception!</code> macro to define a new excetpion type:</p>
<pre><pre class="playpen"><code class="language-rust">
# #![allow(unused_variables)]
#fn main() {
py_exception!(module, MyError);
#}</code></pre></pre>
<ul>
<li><code>module</code> is the name of the containing module.</li>
<li><code>MyError</code> is the name of the new exception type.</li>
</ul>
<p>For example:</p>
<pre><pre class="playpen"><code class="language-rust">#[macro_use] extern crate pyo3;

use pyo3::{Python, PyDict};

py_exception!(mymodule, CustomError);

fn main() {
let gil = Python::acquire_gil();
    let py = gil.python();
    let ctx = PyDict::new(py);

    ctx.set_item(py, &quot;CustomError&quot;, py.get_type::&lt;CustomError&gt;()).unwrap();

    py.run(&quot;assert str(CustomError) == \&quot;&lt;class 'mymodule.CustomError'&gt;\&quot;&quot;, None, Some(&amp;ctx)).unwrap();
    py.run(&quot;assert CustomError('oops').args == ('oops',)&quot;, None, Some(&amp;ctx)).unwrap();
}
</code></pre></pre>
<a class="header" href="./exception.html#raise-an-exception" id="raise-an-exception"><h2>Raise an exception</h2></a>
<p>To raise an exception, first you need to obtain an exception type and construct a new <a href="https://pyo3.github.io/PyO3/pyo3/struct.PyErr.html"><code>PyErr</code></a>, then call <a href="https://pyo3.github.io/PyO3/pyo3/struct.PyErr.html#method.restore"><code>PyErr::restore()</code></a> method to write the exception back to the Python interpreter's global state.</p>
<pre><pre class="playpen"><code class="language-rust">extern crate pyo3;

use pyo3::{Python, PyErr, exc};

fn main() {
    let gil = Python::acquire_gil();
    let py = gil.python();
    PyErr::new::&lt;exc::TypeError, _&gt;(&quot;Error&quot;).restore(py);
    assert!(PyErr::occurred(py));
    drop(PyErr::fetch(py));
}
</code></pre></pre>
<p>If you already have a Python exception instance, you can simply call <a href="https://pyo3.github.io/PyO3/pyo3/struct.PyErr.html#method.from_instance"><code>PyErr::from_instance()</code></a>.</p>
<pre><pre class="playpen"><code class="language-rust">
# #![allow(unused_variables)]
#fn main() {
PyErr::from_instance(py, err).restore(py);
#}</code></pre></pre>
<p>If rust type exists for exception, then it is possible to use <code>new</code> method.
For example each standard exception defined in <code>exc</code> module
has corresponding rust type, exceptions defined by <code>py_exception!</code> and <code>import_exception!</code> macro
have rust type as well.</p>
<pre><pre class="playpen"><code class="language-rust">
# #![allow(unused_variables)]

#fn main() {
fn my_func(arg: PyObject) -&gt; PyResult&lt;()&gt; {
    if check_for_error() {
        Err(exc::ValueError::new(&quot;argument is wrong&quot;))
    } else {
        Ok(())
    }
}

#}</code></pre></pre>
<a class="header" href="./exception.html#check-exception-type" id="check-exception-type"><h2>Check exception type</h2></a>
<p>Python has an <a href="https://docs.python.org/3/library/functions.html#isinstance"><code>isinstance</code></a> method to check object type,
in <code>PyO3</code> there is a <a href="https://pyo3.github.io/PyO3/pyo3/struct.Python.html#method.is_instance"><code>Python::is_instance()</code></a> method which does the same thing.</p>
<pre><pre class="playpen"><code class="language-rust">extern crate pyo3;

use pyo3::{Python, PyBool, PyList};

fn main() {
    let gil = Python::acquire_gil();
    let py = gil.python();
    assert!(py.is_instance::&lt;PyBool&gt;(py.True().as_ref()).unwrap());
    let list = PyList::new(py, &amp;[1, 2, 3, 4]);
    assert!(!py.is_instance::&lt;PyBool&gt;(list.as_ref()).unwrap());
    assert!(py.is_instance::&lt;PyList&gt;(list.as_ref()).unwrap());
}
</code></pre></pre>
<p><a href="https://pyo3.github.io/PyO3/pyo3/struct.Python.html#method.is_instance"><code>Python::is_instance()</code></a> calls the underlaying <a href="https://pyo3.github.io/PyO3/pyo3/struct.PyType.html#method.is_instance"><code>PyType::is_instance</code></a> method to do the actual work.</p>
<p>To check the type of an exception, you can simply do:</p>
<pre><pre class="playpen"><code class="language-rust">
# #![allow(unused_variables)]
#fn main() {
let ret = py.is_instance::&lt;exc::TypeError&gt;(&amp;err.instance(py)).expect(&quot;Error calling is_instance&quot;);
#}</code></pre></pre>
<a class="header" href="./exception.html#handle-rust-error" id="handle-rust-error"><h2>Handle Rust Error</h2></a>
<p>The vast majority of operations in this library will return <a href="https://pyo3.github.io/PyO3/pyo3/type.PyResult.html"><code>PyResult&lt;T&gt;</code></a>.
This is an alias for the type <code>Result&lt;T, PyErr&gt;</code>.</p>
<p>A <a href="https://pyo3.github.io/PyO3/pyo3/struct.PyErr.html"><code>PyErr</code></a> represents a Python exception.
Errors within the <code>PyO3</code> library are also exposed as Python exceptions.</p>
<p>PyO3 library handles python exception in two stages. During first stage <code>PyErr</code> instance get
created. At this stage python gil is not required. During second stage, actual python
exception instance get crated and set to python interpreter.</p>
<p>In simple case, for custom errors support implementation of <code>std::convert::From&lt;T&gt;</code> trait
for this custom error is enough. <code>PyErr::new</code> accepts arguments in form
of <code>ToPyObject + 'static</code>. In case if <code>'static</code> constraint can not be satisfied or
more complex arguments are required <a href="https://pyo3.github.io/PyO3/pyo3/trait.PyErrArguments.html"><code>PyErrArgument</code></a>
trait can be implemented. In that case actual exception arguments creation get delayed
until <code>Python</code> object is available.</p>
<pre><pre class="playpen"><code class="language-rust">
# #![allow(unused_variables)]
#fn main() {
use std;
use std::net::TcpListener;
use pyo3::{PyErr, PyResult, ToPyErr, exc};

impl std::convert::From&lt;std::io::Error&gt; for PyErr {
    fn from(err: std::io::Error) -&gt; PyErr {
        exc::OSError.into()
    }
}

fn connect(s: String) -&gt; PyResult&lt;bool&gt; {
    TcpListener::bind(&quot;127.0.0.1:80&quot;)?;

    Ok(true)
}
#}</code></pre></pre>
<p>The code snippet above will raise <code>OSError</code> in Python if <code>TcpListener::bind()</code> return an error.</p>
<p><code>std::convert::From&lt;T&gt;</code> trait is implemented for most of the standard library's error
types so <code>try!</code> macro or <code>?</code> operator can be used.</p>
<pre><pre class="playpen"><code class="language-rust">
# #![allow(unused_variables)]
#fn main() {
use pyo3::*;

fn parse_int(s: String) -&gt; PyResult&lt;usize&gt; {
    Ok(s.parse::&lt;usize&gt;()?)
}
#}</code></pre></pre>
<p>The code snippet above will raise <code>ValueError</code> in Python if <code>String::parse()</code> return an error.</p>
<a class="header" href="./exception.html#using-exceptions-defined-in-python-code" id="using-exceptions-defined-in-python-code"><h2>Using exceptions defined in python code</h2></a>
<p>It is possible to use exception defined in python code as native rust types.
<code>import_exception!</code> macro allows to import specific exception class and defined zst type
for that exception.</p>
<pre><pre class="playpen"><code class="language-rust">
# #![allow(unused_variables)]
#fn main() {
use pyo3::{PyErr, PyResult, exc};

import_exception!(asyncio, CancelledError)

fn cancel(fut: PyFuture) -&gt; PyResult&lt;()&gt; {
    if fut.cancelled() {
       Err(CancelledError.into())
    }

    Ok(())
}

#}</code></pre></pre>
<p><a href="https://pyo3.github.io/PyO3/pyo3/exc/index.html"><code>exc</code></a> defines exceptions for
several standard library modules.</p>

                </div>

                <!-- Mobile navigation buttons -->
                
                    <a rel="prev" href="./conversions.html" class="mobile-nav-chapters previous">
                        <i class="fa fa-angle-left"></i>
                    </a>
                

                
                    <a rel="next" href="./module.html" class="mobile-nav-chapters next">
                        <i class="fa fa-angle-right"></i>
                    </a>
                

            </div>

            
                <a href="./conversions.html" class="nav-chapters previous" title="You can navigate through the chapters using the arrow keys">
                    <i class="fa fa-angle-left"></i>
                </a>
            

            
                <a href="./module.html" class="nav-chapters next" title="You can navigate through the chapters using the arrow keys">
                    <i class="fa fa-angle-right"></i>
                </a>
            

        </div>


        <!-- Local fallback for Font Awesome -->
        <script>
            if ($(".fa").css("font-family") !== "FontAwesome") {
                $('<link rel="stylesheet" type="text/css" href="_FontAwesome/css/font-awesome.css">').prependTo('head');
            }
        </script>

        <!-- Livereload script (if served using the cli tool) -->
        

        

        

        

        <script src="highlight.js"></script>
        <script src="book.js"></script>
    </body>
</html>
